tools/c7n_azure/tests/test_storage.py

```diff
@@ -42,3 +60,172 @@ def test_value_filter(self):
+                     {'virtual-network-resource-id': id1},
+                     {'virtual-network-resource-id': id2}
+                 ]}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        self.assertEqual(len(resources), 1)
+        rules = resources[0]['properties']['networkAcls']['virtualNetworkRules']
+        self.assertEqual(len(rules), 2)
+        self.assertEqual(rules[0]['id'], id1)
+        self.assertEqual(rules[1]['id'], id2)
+        self.assertEqual(rules[0]['action'], 'Allow')
+        self.assertEqual(rules[1]['action'], 'Allow')
+
+    @arm_template('storage.json')
+    def test_empty_bypass_network_rules_action(self):
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny',
+                 'bypass': []}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        bypass = resources[0]['properties']['networkAcls']['bypass']
+        self.assertEqual(bypass, 'None')
+
+    @arm_template('storage.json')
+    def test_missing_bypass_network_rules_action(self):
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny'}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        bypass = resources[0]['properties']['networkAcls']['bypass']
+        self.assertEqual(bypass, 'None')
+
+    @arm_template('storage.json')
+    def test_bypass_network_rules_action(self):
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny',
+                 'bypass': ['Metrics']}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        bypass = resources[0]['properties']['networkAcls']['bypass']
+        self.assertEqual(bypass, 'Metrics')
+
+    def _get_resources(self):
```

Comment:
why are you running another policy just to get the resources, don't you already have them from the test policy re

```
resources = p_add.run()
```

as its causing a bunch of duplicate recorded json. generally tests that are executing more than one policy are a little suspect on what their testing, and how much extra overhead their adding, we generally construct and use an sdk client to do additional api calls on verification of changes.

Reply:
p_add.run() only runs the action without querying the new state of the resource so the result will be the old state.
I would prefer to use a filter query to get the new state because it is simpler and easier to understand that we want the same resource that we just modified. The number of recorded queries in the cassettes, I think, should be same, regardless of whether we use the sdk directly or wrapped into a policy.

Reply:
typically their is always some recorded json on an extra policy execution (these tests are doing three policies per test) where as doing the api client ensures minimal diff, this pr is weighing at 4k, the majority of which is recorded data, which is always a red flag for extraneous data. its also worthwhile to setup in an environment/storage account that doesn't have .. for other providers the json recorded is actually plain text so in some cases we manually edit it for brevity.

Reply:
alright, using the client to query them